src/command-handler.ts

```diff
@@ -137,19 +137,15 @@ export default class CommandHandler {
 
     if (before != source) {
       await this.activeEditor.edit((edit) => {
-        const firstLine = this.activeEditor!.document.lineAt(0);
-        const lastLine = this.activeEditor!.document.lineAt(
-          this.activeEditor!.document.lineCount - 1
-        );
-
-        const textRange = new vscode.Range(
-          0,
-          firstLine.range.start.character,
-          this.activeEditor!.document.lineCount - 1,
-          lastLine.range.end.character
-        );
-
-        edit.replace(textRange, source);
+        const {
+          start_row,
+          start_column,
+          stop_row,
+          stop_column,
+          insertion_text
+        } = diff.codeDiff(before, source);
+
+        edit.replace(new vscode.Range(start_row, start_column, stop_row, stop_column), insertion_text)
```

Comment:
Thanks for the quick turn-around! It's great to hear this approach fixes the issue for you as well. We're thinking about how to open source more of Serenade, so this is a useful exercise for us as well.

A couple extra thoughts that jump out at me looking at this new implementation:
- It looks like there's actually a function above codeDiff that converts cursor positions to row, column coordinates that we can use.
- Once we have that, we can simplify the signature of codeDiff so that people don't need to jump to the function definition to understand what it does. More specifically, we can change the name to something like beforeRangeChanged.

This code block would end up looking something like this:
```
const { start, stop } = diff.beforeRangeChanged(before, source);
const { startRow, startColumn } = diff.cursorToRowAndColumn(start);
const { stopRow, stopColumn } = diff.cursorToRowAndColumn(start);
let insertionText = before.substring(start, stop);
edit.replace(new vscode.Range(startRow, startColumn, stopRow, stopColumn), insertionText)
```

(also just changed things to camel case for consistency reasons).

Reply:
Open to re-factoring the names to be camel case. My bad on that part.

One of the reasons I did this in one go was that this function as a standalone piece does the minimum required work.

1. We are able to calculate the first cursor position while looking for the first change. This avoids another o(n).
2. Then, when we need the cursor position for the `stop`, we only need to go forward from the `start` and not redo work to calculate from the beginning again. This avoids another o(n). For deletions, this will always be zero and for insertions, this will be the size of the insertion, so this will be small.

I personally prefer the better runtime and fewer allocations version but understand that we are debating over very small differences. So whatever you/serenade decide works with me.

Reply:
I also don’t mind the idea of making the cursor object functional and being able to call “forward“ with text and an amount or something to be able to start from the previous values or something? Would need more thought.